### PR TITLE
[Feature] Support DSVT training

### PR DESCRIPTION
**Pull request by sunjiahao1999**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/models/dense_heads/centerpoint_head.py [hunk 1]
```diff
@@ -101,7 +101,7 @@ def forward(self, x):
         Returns:
             dict[str: torch.Tensor]: contains the following keys:
 
-                -reg （torch.Tensor): 2D regression value with the
```
Comment (sunjiahao1999) [inline mmdet3d/models/dense_heads/centerpoint_head.py:104]: Convert Chinese `（` to English `(`.

File: mmdet3d/models/necks/second_fpn.py [hunk 2]
```diff
@@ -64,12 +74,6 @@ def __init__(self,
             deblocks.append(deblock)
         self.deblocks = nn.ModuleList(deblocks)
 
-        if init_cfg is None:
```
Comment (sunjiahao1999) [inline mmdet3d/models/necks/second_fpn.py:67]: As default init_cfg.

File: mmdet3d/structures/bbox_3d/base_box3d.py [hunk 3]
```diff
@@ -275,12 +275,13 @@ def in_range_3d(
             Tensor: A binary vector indicating whether each point is inside the
             reference range.
         """
-        in_range_flags = ((self.tensor[:, 0] > box_range[0])
```
Comment (sunjiahao1999) [inline mmdet3d/structures/bbox_3d/base_box3d.py:278]: `bbox3d.in_range_3d` should filter out bboxes where the gravity center of the bbox is not in the target range, not the bottom center. Btw, this function not use the training/testing pipeline of any other model, except for DSVT. Therefore, this modification does not cause BC-breaking.

File: tools/train.py [hunk 4]
```diff
@@ -21,6 +21,12 @@ def parse_args():
         action='store_true',
         default=False,
         help='enable automatic-mixed-precision training')
+    parser.add_argument(
```
Comment (sunjiahao1999) [inline tools/train.py:24]: Add `--sync_bn` option in training.

File: projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py [hunk 5]
```diff
@@ -2,6 +2,7 @@
 custom_imports = dict(
     imports=['projects.DSVT.dsvt'], allow_failed_imports=False)
 
+# load_from = 'checkpoints/dsvt_init.pth'
```
Comment (Tai-Wang) [inline projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py:5]: can remove

File: projects/DSVT/dsvt/dynamic_pillar_vfe.py [hunk 6]
```diff
@@ -76,6 +77,8 @@ def __init__(self, with_distance, use_absolute_xyz, use_norm, num_filters,
         self.voxel_x = voxel_size[0]
         self.voxel_y = voxel_size[1]
         self.voxel_z = voxel_size[2]
+        # TODO: remove it after 对齐精度
+        point_cloud_range = np.array(point_cloud_range).astype(np.float32)
```
Comment (Tai-Wang) [inline projects/DSVT/dsvt/dynamic_pillar_vfe.py:81]: can remove

File: projects/DSVT/dsvt/transforms_3d.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
+from typing import List
+
+import numpy as np
+from mmcv import BaseTransform
+
+from mmdet3d.registry import TRANSFORMS
+
+
+@TRANSFORMS.register_module()
+class DSVTObjectRangeFilter(BaseTransform):
```
Comment (Tai-Wang) [inline projects/DSVT/dsvt/transforms_3d.py:10]: Can rename it as ObjectRangeFilter3D and also rename the other functions similarly

File: projects/DSVT/dsvt/utils.py [hunk 8]
```diff
@@ -298,3 +299,141 @@ def decode(self,
...
+@weighted_loss
+def diou3d_loss(pred_boxes, gt_boxes, eps: float = 1e-7):
+    """
+    https://github.com/agent-sgs/PillarNet/blob/master/det3d/core/utils/center_utils.py # noqa
+    Args:
+        pred_boxes (N, 7):
+        gt_boxes (N, 7):
+
+    Returns:
+        _type_: _description_
```
Comment (Tai-Wang) [inline projects/DSVT/dsvt/utils.py:322]: Need to update the docstring

File: projects/DSVT/dsvt/utils.py [hunk 9]
```diff
@@ -298,3 +299,141 @@ def decode(self,
...
+
+    loss = 1 - dious
+
+    return loss
+
+
+@MODELS.register_module()
+class DIoU3DLoss(nn.Module):
+    r"""3D bboxes Implementation of `Distance-IoU Loss: Faster and Better
+    Learning for Bounding Box Regression https://arxiv.org/abs/1911.08287`_.
```
Comment (Tai-Wang) [inline projects/DSVT/dsvt/utils.py:374]: fix the docstring format: <https://arxiv.org/abs/1911.08287>`

File: projects/DSVT/dsvt/utils.py [hunk 10]
```diff
@@ -298,3 +299,141 @@ def decode(self,
...
+
+    Code is modified from https://github.com/Zzh-tju/DIoU.
+
+    Args:
+        eps (float): Epsilon to avoid log(0).
+        reduction (str): Options are "none", "mean" and "sum".
+        loss_weight (float): Weight of loss.
```
Comment (Tai-Wang) [inline projects/DSVT/dsvt/utils.py:381]: Defaults to xxx for all the arguments

File: projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py [hunk 11]
```diff
@@ -212,18 +226,60 @@
...
-    ann_file='./data/waymo/kitti_format/waymo_infos_val.pkl',
     waymo_bin_file='./data/waymo/waymo_format/gt.bin',
-    data_root='./data/waymo/waymo_format',
-    backend_args=backend_args,
-    convert_kitti_format=False,
-    idx2metainfo='./data/waymo/waymo_format/idx2metainfo.pkl')
+    result_prefix='./dsvt_pred')
 test_evaluator = val_evaluator
 
+# vis_backends = [dict(type='LocalVisBackend'), dict(type='WandbVisBackend')]
```
Comment (Tai-Wang) [inline projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py:233]: Can remove

File: projects/DSVT/README.md [hunk 12]
```diff
@@ -57,18 +57,24 @@ python tools/test.py projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-
...
+|                                     Middle Encoder                                     |                                          Backbone                                           | Load Interval | Voxel type (voxel size) | Multi-Class NMS | Multi-frames | mAP@L1 | mAPH@L1 | mAP@L2 | **mAPH@L2** | Download  |
+| :------------------------------------------------------------------------------------: | :-----------------------------------------------------------------------------------------: | :-----------: | :---------------------: | :-------------: | :----------: | :----: | :-----: | :----: | :---------: | :-------: |
+| [DSVT](./configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py) | [ResSECOND](./configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py) |       5       |      voxel (0.32)       |        ✓        |      ×       |  75.5  |  72.4   |  69.2  |    66.3     | [log](<>) |
```
Comment (Tai-Wang) [inline projects/DSVT/README.md:72]: Need to update the log link

File: projects/DSVT/dsvt/dsvt_head.py [hunk 13]
```diff
@@ -79,13 +401,72 @@ def loss_by_feat(self, preds_dicts: Tuple[List[dict]],
                 tasks head, and the internal list indicate different
                 FPN level.
             batch_gt_instances_3d (list[:obj:`InstanceData`]): Batch of
-                gt_instances. It usually includes ``bboxes_3d`` and\
+                gt_instances. It usually includes ``bboxes_3d`` and
```
Comment (ZCMax) [inline projects/DSVT/dsvt/dsvt_head.py:404]: gt_instances_3d

File: projects/DSVT/dsvt/dsvt_head.py [hunk 14]
```diff
@@ -66,7 +98,297 @@ def loss(self, pts_feats: List[Tensor],
...
+                task_box.append(gt_bboxes_3d[m])
+                # 0 is background for each task, so we need to add 1 here.
+                task_class.append(gt_labels_3d[m] + 1 - flag2)
+            task_boxes.append(torch.cat(task_box, axis=0).to(device))
+            task_classes.append(torch.cat(task_class).long().to(device))
+            flag2 += len(mask)
+        draw_gaussian = draw_heatmap_gaussian
+        heatmaps, anno_boxes, inds, masks = [], [], [], []
+
+        for idx, task_head in enumerate(self.task_heads):
```
Comment (ZCMax) [inline projects/DSVT/dsvt/dsvt_head.py:313]: some basic comments are needed to introduce the target generation process

File: projects/DSVT/dsvt/dynamic_pillar_vfe.py [hunk 15]
```diff
@@ -76,6 +77,8 @@ def __init__(self, with_distance, use_absolute_xyz, use_norm, num_filters,
         self.voxel_x = voxel_size[0]
         self.voxel_y = voxel_size[1]
         self.voxel_z = voxel_size[2]
+        # TODO: remove it after 对齐精度
```
Comment (ZCMax) [inline projects/DSVT/dsvt/dynamic_pillar_vfe.py:80]: Clear the comments?

File: projects/DSVT/dsvt/utils.py [hunk 16]
```diff
@@ -298,3 +299,141 @@ def decode(self,
...
+        device=dim.device).type_as(center)  # (4, 2)
+    corners = dim.view([-1, 1, 2]) * corners_norm.view([1, 4, 2])  # (N, 4, 2)
+    corners = corners + center.view(-1, 1, 2)
+    return corners
+
+
+@weighted_loss
+def diou3d_loss(pred_boxes, gt_boxes, eps: float = 1e-7):
+    """
+    https://github.com/agent-sgs/PillarNet/blob/master/det3d/core/utils/center_utils.py # noqa
```
Comment (ZCMax) [inline projects/DSVT/dsvt/utils.py:316]: modified from xxxx

File: projects/DSVT/dsvt/res_second.py [hunk 17]
```diff
@@ -78,17 +76,14 @@ class ResSECOND(BaseModule):
         out_channels (list[int]): Output channels for multi-scale feature maps.
         blocks_nums (list[int]): Number of blocks in each stage.
         layer_strides (list[int]): Strides of each stage.
-        norm_cfg (dict): Config dict of normalization layers.
-        conv_cfg (dict): Config dict of convolutional layers.
```
Comment (ZCMax) [inline projects/DSVT/dsvt/res_second.py:82]: lack init_cfg docstrings

File: projects/DSVT/dsvt/dsvt_head.py [hunk 18]
```diff
@@ -18,8 +25,33 @@ class DSVTCenterHead(CenterHead):
     This head adds IoU prediction branch based on the original CenterHead.
     """
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self,
+                 *args,
```
Comment (ZCMax) [inline projects/DSVT/dsvt/dsvt_head.py:29]: better to use (loss_iou, loss_reg_iou, *args, **kwargs)

File: projects/DSVT/dsvt/dsvt_head.py [hunk 19]
```diff
@@ -66,7 +98,297 @@ def loss(self, pts_feats: List[Tensor],
...
+    def calc_iou_loss(self, iou_preds, batch_box_preds, mask, ind, gt_boxes):
+        """
+        Args:
+            iou_preds: (batch x 1 x h x w)
+            batch_box_preds: (batch x (7 or 9) x h x w)
+            mask: (batch x max_objects)
+            ind: (batch x max_objects)
+            gt_boxes: List of batch groundtruth boxes.
+
+        Returns:
```
Comment (ZCMax) [inline projects/DSVT/dsvt/dsvt_head.py:167]: Returns: Loss tensor

File: projects/DSVT/dsvt/dsvt_head.py [hunk 20]
```diff
@@ -66,7 +98,297 @@ def loss(self, pts_feats: List[Tensor],
...
+        # Transpose task_gt_bboxes
+        task_gt_bboxes = list(map(list, zip(*task_gt_bboxes)))
+        return heatmaps, anno_boxes, inds, masks, task_gt_bboxes
+
+    def get_targets_single(self,
+                           gt_instances_3d: InstanceData) -> Tuple[Tensor]:
+        """Generate training targets for a single sample.
+
+        Args:
+            gt_instances_3d (:obj:`InstanceData`): Gt_instances of
```
Comment (ZCMax) [inline projects/DSVT/dsvt/dsvt_head.py:259]: Gt_instances_3d

File: projects/DSVT/README.md [hunk 21]
```diff
@@ -57,18 +57,24 @@ python tools/test.py projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-
...
+| [DSVT](./configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py) | [ResSECOND](./configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py) |       5       |      voxel (0.32)       |        ✓        |      ×       |  75.5  |  72.4   |  69.2  |    66.3     | \[log\](\<https://download.openmmlab.com/mmdetection3d/v1.1.0_models/dsvt/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class_20230917_102130.log) |
 
 **Note** that `ResSECOND` denotes the base block in SECOND has residual layers.
 
+**Note** Regrettably, we are unable to provide the pre-trained model weights due to [Waymo Dataset License Agreement](https://waymo.com/open/terms/). However, we can provide the training logs.
```
Comment (Tai-Wang) [inline projects/DSVT/README.md:76]: **Note**:
-  that `ResSECOND` denotes the base block in SECOND has residual layers.
- Regrettably, we are unable to provide the pre-trained model weights due to [Waymo Dataset License Agreement](https://waymo.com/open/terms/), so we only provide the training logs as shown above.

File: projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py [hunk 22]
```diff
@@ -172,25 +175,35 @@
...
+        keys=['points'],
+        meta_keys=['box_type_3d', 'sample_idx', 'context_name', 'timestamp'])
 ]
 
 dataset_type = 'WaymoDataset'
+train_dataloader = dict(
+    batch_size=1,
+    num_workers=4,
+    persistent_workers=True,
+    # sampler=dict(type='DefaultSampler', shuffle=False),
```
Comment (Tai-Wang) [inline projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py:190]: remove unused codes

File: projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py [hunk 23]
```diff
@@ -138,25 +143,23 @@
...
         rot_range=[-0.78539816, 0.78539816],
         scale_ratio_range=[0.95, 1.05],
-        translation_std=[0.5, 0.5, 0]),
-    dict(type='PointsRangeFilter', point_cloud_range=point_cloud_range),
-    dict(type='ObjectRangeFilter', point_cloud_range=point_cloud_range),
-    dict(type='ObjectNameFilter', classes=class_names),
+        translation_std=[0.5, 0.5, 0.5]),
+    dict(type='PointsRangeFilter3D', point_cloud_range=point_cloud_range),
+    dict(type='ObjectRangeFilter3D', point_cloud_range=point_cloud_range),
+    # dict(type='ObjectNameFilter', classes=class_names),
```
Comment (Tai-Wang) [inline projects/DSVT/configs/dsvt_voxel032_res-second_secfpn_8xb1-cyclic-12e_waymoD5-3d-3class.py:162]: can remove

File: projects/DSVT/dsvt/utils.py [hunk 24]
```diff
@@ -298,3 +299,142 @@ def decode(self,
...
+        device=dim.device).type_as(center)  # (4, 2)
+    corners = dim.view([-1, 1, 2]) * corners_norm.view([1, 4, 2])  # (N, 4, 2)
+    corners = corners + center.view(-1, 1, 2)
+    return corners
+
+
+@weighted_loss
+def diou3d_loss(pred_boxes, gt_boxes, eps: float = 1e-7):
+    """
+    modified from https://github.com/agent-sgs/PillarNet/blob/master/det3d/core/utils/center_utils.py # noqa
```
Comment (Tai-Wang) [inline projects/DSVT/dsvt/utils.py:316]: rephrase the docstring

File: projects/DSVT/dsvt/utils.py [hunk 25]
```diff
@@ -298,3 +299,142 @@ def decode(self,
...
+
+    loss = 1 - dious
+
+    return loss
+
+
+@MODELS.register_module()
+class DIoU3DLoss(nn.Module):
+    r"""3D bboxes Implementation of `Distance-IoU Loss: Faster and Better
+    Learning for Bounding Box Regression <https://arxiv.org/abs/1911.08287>`.
```
Comment (Tai-Wang) [inline projects/DSVT/dsvt/utils.py:374]: \<https://arxiv.org/abs/1911.08287>`_.

**Review by Tai-Wang — APPROVED** [review]
